python/ray/rllib/agent.py

```diff
@@ -63,7 +63,7 @@ class Agent(Trainable):
     _allow_unknown_subkeys = []
 
     def __init__(
-            self, config={}, env=None, registry=get_registry(),
+            self, config=None, env=None, registry=None,
```

Comment:
We should keep it at get_registry(), otherwise you can't use custom envs by default.

Reply:
This will do in `Trainable` class init method, I do not know if I understand the right?

![image](https://user-images.githubusercontent.com/7162889/35722633-0c77fe9a-0832-11e8-9f8e-05be2699d441.png)

Reply:
Forgive me, my English is not so good， I wish I could explain this problem well.

No mater how many instances of `Agent` class were created, the `get_register()` method will call only once by default. Running the code like below, the console will output `get_registry calling`  only once.

``` python
def get_register():
    print("get_registry calling")
    return {}

class A(object):
    def __init__(self, register=get_register()):
        self.register = register

a = A()
b = A()
```

Notice the fact that when running the unit tests in `test_supported_spaces.py`, the `get_registry` method will be called before the `registry_env` in `check_support` method. The calling stacktrace like below:

![image](https://user-images.githubusercontent.com/7162889/35725927-186f3a2c-083e-11e8-9f4a-c333413c3236.png)


Before this "PullReqeust", whether reigstry created by `get_registry` or the `_default_registry` will share the same dict object named `_all_objects` in `_Registry` class, therefore, even we call `register_env` after `get_registry`, we can also read `stub_env`.

But in this "PullRequest", we do dict copy when creating `_Registry` object, so the `_all_objects` will not share between instances of `_Registry`, so unit test failed for `malformed environment ID: stub_env`.

![image](https://user-images.githubusercontent.com/7162889/35726177-df617ae6-083e-11e8-89e5-0b6d2f5988f6.png)

Reply:
@ericl what about adding a line like

```python
if registry is None:
    registry = get_registry()
```

inside the `__init__` function. Would that address the issue?

Reply:
@the-sea if the registry issue is more complex we can remove it from this PR, merge the other changes, and look into the registry issue in a follow-up PR.

Reply:
Let me look into this -- there might be some latent bug this uncovered.

Reply:
@robertnishihara  Ok, I will remove changes about registry.

Reply:
@the-sea I've just tried it out and the test_supported_spaces does seem to pass if you set this to None. Thanks for the explanation, this looks good to me!

Reply:
@ericl Thank you, Do I need to create another PR to fix the registry issue?

Reply:
That would be great.